### ISSUE TITLE
sendMessage aborted

### ISSUE THREAD

**Issue by averlon** [Robertof/perl-www-telegram-botapi#33]
Hi,
the docu says:
`
If the request is not successful or the server tells us something isn't ok, then this method dies with the first available error message (either the error description or the status line). You can make this method non-fatal using eval:

my $response = eval { $api->api_request ($method, $args) }
    or warn "Request failed with error '$@', but I'm still alive!";
`

I have used:
`
$av_std_RETVAL = eval 
{
  $av_obj_TGRAM->sendMessage (
    {
        chat_id => $av_loc_tgram_CHATID,
        text    => $av_tmp_STRING,
        disable_notification => 'false',
        parse_mode => 'HTML',
    }
  )
}
or warn "Request failed with error '$@', but I'm still alive!";
`

The string contains unsuported formatting tags:
`
Logfile: <strong>my test logfile</strong> xxx<pre>xxx<www>xxx<strong>xxx
`
Error:
`
Request failed with error 'ERROR: code 400: Bad Request: can't parse entities: Unsupported start tag "www" at byte offset 53 at /usr/share/perl5/WWW/Telegram/BotAPI.pm line 224.
        WWW::Telegram::BotAPI::api_request(undef, undef, HASH(0x55e57ea8da98)) called at ./av_perlre.pl line 268
        eval {...} called at ./av_perlre.pl line 266
', but I'm still alive! at ./av_perlre.pl line 266.
`
Is there any way to let the script still run and grab some error message?

**Comment by Robertof** [comment]
Hello! Thanks for your report.

Could you clarify what problem do you have? Is the problem that your program still exits when the API returns an error even if you have the `eval` block, or that you would like to extract the error message?

If it's the former, it's weird since `eval` should trap the errors (and in fact yours looks like a warning, not a fatal). If it's the latter, you should be able to use [`parse_error`](https://github.com/Robertof/perl-www-telegram-botapi#parse_error) to get more information about the message.

Let me know!

**Comment by averlon** [comment]
Hi,
I expected the program by using the eval to continue - but it aborts.
I am not a very experience programmer. Probably I have done something wrong on the eval statement.
But probably eval simply does not work in this environment - I don't know!

Probably you could do something inside your code that it is not aborted. I don't know!

From my perspective, I have done it according to the docu. And this stated, that the program will then continue.

**Comment by Robertof** [comment]
Hey,

thanks for clarifying! This sounds very strange. The docs for [`eval`](https://perldoc.perl.org/functions/eval) are pretty explicative on the fact that they trap fatal errors that happen inside the block. Any chance you have a custom `$SIG{'__DIE__'}` that might interfere with it? If so, you could try with:

```
eval { local $SIG{'__DIE__'}; $av_obj_TGRAM->sendMessage(...) }
```

Or maybe any chance you have made warnings fatal e.g. with `use warnings FATAL => 'all'`? That would also not work correctly with your code.

The last thing I can think of is that maybe you're doing something with `$av_std_RETVAL` when it fails even if it's undefined.

It would be helpful if you could share more of your code to figure more out!

**Comment by averlon** [comment]
Hi,
neither, nor. At least, not that I am aware of.
I have attached a testscript which I have used to play arround with the error.
It will not run, since the TELEGRAM settings are not complete. But I am sure you have own tokens and chat-IDs.
[av_perlre_github.txt](https://github.com/Robertof/perl-www-telegram-botapi/files/11645797/av_perlre_github.txt)

**Comment by Robertof** [comment]
Thanks for attaching the script! I gave it a spin without even changing the token (which errors out). I first ran the script as-is:

```
$ perl av_perlre_github.pl
alle optionen: 
xxx<pre>xxx
<
www>xxx<strong>xxx
Request failed with error 'ERROR: code 404: Not Found at /opt/homebrew/opt/perl/lib/perl5/site_perl/5.36/WWW/Telegram/BotAPI.pm line 224.
	WWW::Telegram::BotAPI::api_request(undef, undef, HASH(0x127c58a60)) called at /opt/homebrew/opt/perl/lib/perl5/site_perl/5.36/WWW/Telegram/BotAPI.pm line 92
	WWW::Telegram::BotAPI::AUTOLOAD(WWW::Telegram::BotAPI=HASH(0x127971170), HASH(0x127c58a60)) called at av_perlre_github.pl line 267
	eval {...} called at av_perlre_github.pl line 266
', but I'm still alive! at av_perlre_github.pl line 266.
$ echo $?
0
```

This doesn't tell us whether the script exited or not though, since there's nothing after the `eval` block. I tried to add this line after `or warn "Request..."`:

```perl
print "Testing 123... if the script has died this should not print\n";
```

And I got this:
```
alle optionen: 
xxx<pre>xxx
<
www>xxx<strong>xxx
Request failed with error 'ERROR: code 404: Not Found at /opt/homebrew/opt/perl/lib/perl5/site_perl/5.36/WWW/Telegram/BotAPI.pm line 224.
	WWW::Telegram::BotAPI::api_request(undef, undef, HASH(0x144832278)) called at /opt/homebrew/opt/perl/lib/perl5/site_perl/5.36/WWW/Telegram/BotAPI.pm line 92
	WWW::Telegram::BotAPI::AUTOLOAD(WWW::Telegram::BotAPI=HASH(0x144a0df70), HASH(0x144832278)) called at av_perlre_github.pl line 267
	eval {...} called at av_perlre_github.pl line 266
', but I'm still alive! at av_perlre_github.pl line 266.
Testing 123... if the script has died this should not print
```

Which seems to imply that `eval` works here and is not killing the script!

To give it one last try I changed your invocation like this:

```perl
$av_std_RETVAL="";
unless (
  $av_std_RETVAL = eval { $av_obj_TGRAM->sendMessage ({
      chat_id => $av_loc_tgram_CHATID,
      text    => $av_tmp_STRING,
      disable_notification => 'false',
      parse_mode => 'HTML',
  }) }
) {
  my $err = $av_obj_TGRAM->parse_error;
  print "Oops! The request failed: ", Data::Dumper::Dumper($err), "\n";
}

print "... but the script is still alive!\n";
```

and I got this output:

```
alle optionen: 
xxx<pre>xxx
<
www>xxx<strong>xxx
Oops! The request failed: $VAR1 = {
          'msg' => 'Not Found',
          'code' => '404',
          'type' => 'api'
        };

... but the script is still alive!
```

Which again suggests that `eval` is doing its job here and *not* killing the script! Do you get a different behavior here?

**Comment by averlon** [comment]
Hi Roberto,
thanks for your testing!
My fault!

I ran the test-script in debug mode. The debug stopped and I misunderstood this as abortion of the script.
I now ran the test-script without debug and you are right, the script survived.

The production script had still a "die" instead of a "warn". I now changed this.

But anyhow I need to find a way to catch the error and eliminate the unsupported formatting string . But this is my problem.

I will close the ticket!
